p2p/starknet/handlers.go

```diff
@@ -101,75 +101,118 @@ func (h *Handler) reqHandler(req *spec.Request) (Stream[proto.Message], error) {
+			// todo log an error
+			return nil, false
+		}
+		it.Next()
+
+		// todo fill the spec later
+		return &spec.BlockBodiesResponse{
+			BlockNumber: block.Number,
+		}, false
 	}, nil
 }
 
-func (h *Handler) HandleGetSignatures(req *spec.GetSignatures) (*spec.Signatures, error) {
+func (h *Handler) HandleGetSignatures(req *spec.Iteration) (*spec.Signatures, error) {
 	// todo: read from bcReader and adapt to p2p type
 	return &spec.Signatures{
-		Id: req.Id,
+		// Id: req.Id,
 	}, nil
 }
 
-func (h *Handler) HandleGetEvents(req *spec.GetEvents) (*spec.Events, error) {
-	block, err := h.blockByID(req.Id)
+func (h *Handler) HandleGetEvents(req *spec.Iteration) (*spec.EventsResponse, error) {
+	block, err := h.blockByIteration(req)
 	if err != nil {
 		return nil, err
 	}
 
-	var result spec.Events
+	var events []*spec.Event
 	for _, receipt := range block.Receipts {
 		for _, ev := range receipt.Events {
 			event := &spec.Event{
 				FromAddress: core2p2p.AdaptFelt(ev.From),
-				Keys:        utils.Map(ev.Keys, core2p2p.AdaptFelt),
-				Data:        utils.Map(ev.Data, core2p2p.AdaptFelt),
+				Keys:        core2p2p.AdaptFeltSlice(ev.Keys),
+				Data:        core2p2p.AdaptFeltSlice(ev.Data),
 			}
 
-			result.Events = append(result.Events, event)
+			events = append(events, event)
 		}
 	}
 
-	return &result, nil
+	return &spec.EventsResponse{
+		BlockNumber: block.Number,
+		BlockHash:   core2p2p.AdaptHash(block.Hash),
+		Responses: &spec.EventsResponse_Events_{
+			Events: &spec.EventsResponse_Events{
+				Items: events,
+			},
+		},
+	}, nil
 }
 
-func (h *Handler) HandleGetReceipts(req *spec.GetReceipts) (*spec.Receipts, error) {
+func (h *Handler) HandleGetReceipts(req *spec.Iteration) (*spec.ReceiptsResponse, error) {
 	// todo: read from bcReader and adapt to p2p type
-	magic := 37
-	return &spec.Receipts{
-		Receipts: make([]*spec.Receipt, magic),
+	magic := uint64(37) //nolint:gomnd
+	return &spec.ReceiptsResponse{
+		BlockNumber: magic,
+		BlockHash:   nil,
+		Responses:   nil,
 	}, nil
 }
 
-func (h *Handler) HandleGetTransactions(req *spec.GetTransactions) (*spec.Transactions, error) {
-	// todo: read from bcReader and adapt to p2p type
-	magic := 1337
-	return &spec.Transactions{
-		Transactions: make([]*spec.Transaction, magic),
+func (h *Handler) HandleGetTransactions(i *spec.Iteration) (Stream[proto.Message], error) {
+	it, err := h.blockIterator(i)
+	if err != nil {
+		return nil, err
+	}
+
+	return func() (proto.Message, bool) {
+		if !it.Valid() {
+			return nil, false
+		}
+
+		block := it.Block()
+		if it.Err() != nil {
+			h.log.Errorw("Iterator failure", "err", it.Err())
+			return nil, false
+		}
+		it.Next()
+
+		return &spec.TransactionsResponse{
+			BlockNumber: block.Number,
+			BlockHash:   core2p2p.AdaptHash(block.Hash),
+			Responses: &spec.TransactionsResponse_Transactions{
+				Transactions: &spec.Transactions{
+					Items: utils.Map(block.Transactions, core2p2p.AdaptTransaction),
+				},
+			},
+		}, false
```

Comment:
Iterator implementation looks good to me overall.

Should we return `true` here, and return `false` only when `it.Err() != nil` or `!it.Valid()`?